aspnetcore/fundamentals/servers/kestrel.md

```diff
@@ -2707,6 +2707,14 @@ Host Filtering Middleware is disabled by default. To enable the middleware, defi
 
 ::: moniker-end
 
+## Request draining
+
+Since opening a http connection is relatively expensive, Kestrel tries to re-use connections when possible. However, sometimes a request may not be fully consumed, for example if a POST request is made but the server returns a redirect, the client may already have sent part of the post data as the server writes the 301 response. At this point, the connection cannot be used for a new request since there is data from the old request still waiting to be read. For this reason, kestrel tries to drain the request, that is, read and discard the data waiting to be read. This process has a timeout of 5 seconds, if all the data (specified by the content-length header) has not been read at that point, the connection is closed.
+
+While this is a good trade off in most cases there are times where an application may want to terminate the request immediately after writing the response. For example, the clients may have restrictive data caps and limiting uploaded data might be a priority. In those cases, the application may call `HttpContext.Abort()` from either a controller or a custom middleware. However, there are caveats to this. Besides it being expensive to close and open tcp connections, there is no guarantee that the client has actually read the response  before the connection closes. Calling `HttpResponse.CompleteAsync()` before calling `HttpContext.Abort()` will at least ensure that the response has left the server, but client behavior is still not completely predictable.
+
+This approach should therefore only be used when a specific problem needs to be solved, such as if malicious clients are trying to post garbage data or when there is a bug in client code that causes large or numerous requests. If possible, it's better for clients to utilize the 100 continue http response code, so that they wait for the server to respond before starting to send the request body.
```

Comment:
```suggestion
This approach should therefore only be used when a specific problem needs to be solved, such as if malicious clients are trying to post garbage data or when there is a bug in client code that causes large or numerous requests. If possible, it's better for clients to utilize the `Expect: 100-continue` request header so that they wait for the server to respond before starting to send the request body.
```